src/lib/mina/actions/offchain-contract.unit-test.ts

```diff
@@ -20,80 +22,99 @@ const offchainState = OffchainState(
 
 class StateProof extends offchainState.Proof {}
 
-// example contract that interacts with offchain state
+function ExampleContractWithState() {
+  const offchainStateInstance = offchainState.init();
 
-class ExampleContract extends SmartContract {
-  @state(OffchainState.Commitments) offchainState = offchainState.commitments();
+  // example contract that interacts with offchain state
+  class ExampleContract extends SmartContract {
+    @state(OffchainState.Commitments) offchainStateCommitments =
+      offchainState.emptyCommitments();
 
-  @method
-  async createAccount(address: PublicKey, amountToMint: UInt64) {
-    // setting `from` to `undefined` means that the account must not exist yet
-    offchainState.fields.accounts.update(address, {
-      from: undefined,
-      to: amountToMint,
-    });
+    // Since this line is executed over and over, every time a proof is generated, we can't rely on setting this value elsewhere
+    // We also can't rely on different instances of the contract having different offchain state instances, because this specific instance of
+    // offchain state is referenced by the class definition of the smart contract.
+    // By using a closure and exporting the offchain state and the contract together, we can always refer to the correct instance of the offchain state
+    offchainState = offchainStateInstance;
```

Comment:
This line is the source of a lot of frustration.  Here are some alternatives attempted:

- `offchainState = offchainState.init()`
    - Re-initialized the state every time the prover is invoked.  The smart contract is not capable of staying up to date with the contract state.
- `offchainState: typeof offchainStateInstance`
    - The contract can't compile because provable methods rely on offchainState being defined at compile time


That leaves the implementation as implemented here.  The instance `offchainStateInstance` is tied to the class definition of the smart contract.  That means 2 or more smart contract instances will all have the same offchain state instance.  Even if we reset the instance state after initialization, the prover will re-run this line and set the offchain state back to the orifinally-defined value.

I found the best/only way to manage this is to turn the whole smart contract class into a closure so that the definition of each smart contract can be paired with its offchain state instance.  This doesn't feel amazing, but it does unblock the use case.